tests/quick.rs

```diff
@@ -1,12 +1,129 @@
+        );
+
+        assert!(
+            $output.stdout.is_empty(),
+            "stdout: {:?}",
+            String::from_utf8_lossy(&$output.stdout)
+        );
+
+        assert_eq!(
+            String::from_utf8_lossy(&$output.stderr).replace(bzip2_binary(), "bzip2"),
+            $expected_stderr,
+        );
+    };
+}
+
+#[macro_export]
+macro_rules! expect_failure {
+    ($cmd:expr, $expected_stderr:expr $(,)?) => {
+        let cmd = $cmd;
+        let output = match cmd.output() {
+            Ok(output) => output,
+            Err(err) => panic!("Running {:?} failed with {err:?}", cmd),
+        };
+
+        expect_output_failure!(output, $expected_stderr);
+    };
+}
+
+#[macro_export]
+macro_rules! expect_output_success {
+    ($output:expr, $expected_stderr:expr $(,)?) => {
+        assert!(
+            $output.status.success(),
+            "status is failure (expected success) {:?} stderr: {:?}",
+            $output.status,
+            String::from_utf8_lossy(&$output.stderr)
+        );
+
+        assert_eq!(
+            String::from_utf8_lossy(&$output.stderr).replace(bzip2_binary(), "bzip2"),
+            $expected_stderr,
+        );
+    };
+}
+
+#[macro_export]
+macro_rules! expect_success {
+    ($cmd:expr, $expected_stderr:expr $(,)?) => {
+        let cmd = $cmd;
+        let output = match cmd.output() {
+            Ok(output) => output,
+            Err(err) => panic!("Running {:?} failed with {err:?}", cmd),
+        };
+
+        expect_output_success!(output, $expected_stderr);
+    };
+}
+
+#[cfg(unix)]
+unsafe fn custom_tty() -> (
+    std::sync::MutexGuard<'static, ()>,
+    std::process::Stdio,
+    std::process::Stdio,
+) {
+    use std::{os::fd::FromRawFd, sync::Mutex};
+
+    // only one thread should be doing this at a time
+    static X: Mutex<()> = Mutex::new(());
+
+    let guard = X.lock().unwrap();
+
+    // Open the master PTY using libc
+    let cstring = std::ffi::CString::new("/dev/ptmx").unwrap();
+    let master_fd = unsafe { libc::open(cstring.as_ptr(), libc::O_RDWR) };
+    if master_fd < 0 {
+        panic!("{}", std::io::Error::last_os_error());
+    }
+
+    // Grant access to the slave PTY
+    if unsafe { libc::grantpt(master_fd) } < 0 {
+        panic!("{}", std::io::Error::last_os_error());
+    }
+
+    // Unlock the slave PTY
+    if unsafe { libc::unlockpt(master_fd) } < 0 {
+        panic!("{}", std::io::Error::last_os_error());
+    }
+
+    // Get the path to the slave PTY
+    let slave_path_ptr = unsafe { libc::ptsname(master_fd) };
+    if slave_path_ptr.is_null() {
+        panic!("{}", std::io::Error::last_os_error());
+    }
+
+    // Open the slave PTY
+    let slave_fd = unsafe { libc::open(slave_path_ptr, libc::O_RDWR) };
+    if slave_fd < 0 {
+        panic!("{}", std::io::Error::last_os_error());
+    }
+
+    unsafe {
+        (
+            guard,
+            Stdio::from_raw_fd(master_fd),
+            Stdio::from_raw_fd(slave_fd),
+        )
+    }
+}
```

Comment:
the internet told me this is how to set up a TTY. it appears to work well as a stdout device, but not for stdin. It does work locally but that might be because of my shell? anyway on CI it does not work.

Reply:
I've removed the TTY stuff for now, the other changes are more important